option-bucket/src/lib.rs

```diff
@@ -0,0 +1,380 @@
+//!     match OptionBucket::new(this) {
+//!         OptionBucket::Some(inner) => Entry::Occupied(OccupiedEntry { inner }),
+//!         OptionBucket::None(inner) => Entry::Vacant(VacantEntry { inner }),
+//!     }
+//! }
+//! ```
+//!
+//! # Unsafe Usage
+//!
+//! This crate uses `unwrap_unchecked` to obtain
+//! mutable references to the inner value of `Some`s,
+//! and to skip `drop` when overwriting `None`s.
+
+#![no_std]
+#![deny(missing_docs)]
+// Enable pedantic lints as warnings so we don't break builds when
+// lints are modified or new lints are added to clippy.
+#![warn(
+    // Enable more useful rustc lints
+    absolute_paths_not_starting_with_crate,
+    elided_lifetimes_in_paths,
+    explicit_outlives_requirements,
+    keyword_idents,
+    macro_use_extern_crate,
+    meta_variable_misuse,
+    missing_copy_implementations,
+    missing_docs,
+    non_ascii_idents,
+    noop_method_call,
+    pointer_structural_match,
+    single_use_lifetimes,
+    trivial_casts,
+    trivial_numeric_casts,
+    unreachable_pub,
+    unused_extern_crates,
+    unused_import_braces,
+    unused_lifetimes,
+    unused_macro_rules,
+    unused_qualifications,
+    unused_tuple_struct_fields,
+    variant_size_differences,
+    // Enable pedantic clippy lints
+    clippy::pedantic,
+    // Useful clippy lints for no_std support
+    clippy::std_instead_of_core,
+    clippy::std_instead_of_alloc,
+    clippy::alloc_instead_of_core,
+    // Useful clippy lint for unsafe code
+    clippy::undocumented_unsafe_blocks,
+    // Always require `unsafe` block when using `unsafe`
+    // operations, even in `unsafe` functions
+    unsafe_op_in_unsafe_fn,
+)]
+// `clippy::pedantic` exceptions
+#![allow(clippy::should_implement_trait, clippy::must_use_candidate)]
+
+/// Abstraction for an [`&mut Option`][Option] that's known to be [`Some`].
+///
+/// # Size
+///
+/// `SomeOption` is the size of two pointers, making it
+/// twice the size of `&mut Option`. One points to the
+/// value inside, and the other points to the `Option` itself.
+pub struct SomeBucket<'a, T> {
+    outer: &'a mut Option<T>,
+}
+impl<'a, T> SomeBucket<'a, T> {
+    /// Creates a new [`SomeBucket`], without checking that
+    /// the input [`Option`] is `Some`.
+    ///
+    /// It's recommended to use [`SomeBucket::new`] or
+    /// [`OptionBucket::new`] instead.
+    ///
+    /// # Safety
+    ///
+    /// Caller must guarantee that `opt` is NOT `None`.
+    pub unsafe fn new_unchecked(opt: &'a mut Option<T>) -> Self {
+        debug_assert!(
+            opt.is_some(),
+            "Undefined Behavior: `None` value passed to `SomeBucket::new_unchecked`."
+        );
+
+        SomeBucket { outer: opt }
+    }
+
+    /// Creates a new [`SomeBucket`]. Returns `Some(SomeBucket<T>)`
+    /// if `opt` is [`Some`], otherwise returns `None`.
+    ///
+    /// For an unchecked version, see [`SomeBucket::new_unchecked`].
+    pub fn new(opt: &'a mut Option<T>) -> Option<Self> {
```

Comment:
Another thing I realized you probably want to do is that since the `{None,Some,Option}Storage` are used in `fixed-map-derive`, most if its public API (unrelated to trait impls) should probably be `#[doc(hidden)]` and documented as having no stability guarantees.

Reply:
`option_bucket` is `doc(hidden)` and I've added a warning to the module documentation.